### ISSUE TITLE
Allow transient and singleton deps in parent class w/o marking parent class as transient

### ISSUE THREAD

**Issue by angelhdzdev** [robak86/hardwired#87]
👋 hello, nice library! 🤯

Have a question, how to allow parent class to have singleton and transient deps, without marking the parent class as transient?

```ts
class B {
   static class = cls.transient(this)
   count = 0
   callB() { 
     console.log(`count ${this.count}`)
     this.count++
   }
}

class A {
static class = cls.singleton(this, [B.class])
constructor(private readonly b: B) {}

callB() {
this.b.callB()
}
}

let a = container.use(A.class)
a.callB()
a.callB()
a.callB()

a = container.use(A.class)
a.callB()
a.callB()
a.callB()
```

Class A should be singleton, but class B should be transient, so every time a new instance of A is created, B is still reused.

I got it working if I use `static class = cls.transient(this, [B.class])` in A.
But what if I have other deps, lets say B is transient, but C and D are singleton?

Hope you can help thanks.

**Comment by angelhdzdev** [comment]
I tested the hypothetical scenario of having C and D deps and yes, the issue is still present.
Making `A` singleton makes all deps singleton as well, even if they are marked as transient in their own class definitions.
So to make it work I have to set `A` to `transient`:

```ts


class B {
  static class = cls.transient(this)
  count = 0

  b() {
    console.log(`B Count ${this.count}`)
    this.count++
  }
}

class C {
  static class = cls.singleton(this)
  count = 0

  c() {
    console.log(`C Count ${this.count}`)
    this.count++
  }
}

class A {
  static class = cls.transient(this, [B.class, C.class])  // Using transient here gives the impression that all deps are transient, only "B" is transient.

  constructor(private b: B, private c: C) {}

  call() {
    this.b.b()
    this.c.c()
  }
}

let a = container.use(A.class)
a.call()
a.call()
a.call()
a = container.use(A.class)
a.call()
a.call()
a.call()
```

Or maybe I'm misunderstanding the logic.

**Comment by robak86** [comment]
Hey, thanks for your interest in the library!  You’ve just created the first issue in the repo! 🎉

I’ve modified your code slightly to demonstrate the behavior you’re asking about, by adding messages to each constructor to track when instances are created.

```typescript
class BTransient {
  static class = cls.transient(this);

  constructor() {
    console.log('BTransient created');
  }
}

class CSingleton {
  static class = cls.singleton(this);
  count = 0;

  constructor() {
    console.log('CSingleton created');
  }
}

class ATransient {
  static class = cls.transient(this, [BTransient.class, CSingleton.class]);

  constructor(
    private b: BTransient,
    private c: CSingleton,
  ) {
    console.log('ATransient created');
  }
}

class ASingleton {
  static class = cls.singleton(this, [BTransient.class, CSingleton.class]);

  constructor(
    private b: BTransient,
    private c: CSingleton,
  ) {
    console.log('ASingleton created');
  }
}

```
First, let's create a new container:

```
const cnt = container.new();
```

Although you could use `container.use()`, it's a global container. For this example, we'll ensure a clean state by using a new container instance.


Now, when you call for the first time `cnt.use(ATransient.class);` you will see
```
BTransient created
CSingleton created
ATransient created
```
Since the container doesn't have any memoized instances yet, it creates the entire dependency graph from scratch.

2. When you call for the second time `cnt.use(ATransient.class);` you will see
```
BTransient created
ATransient created
```
Here, the transient dependencies (`BTransient` and `ATransient`) are recreated. However, `CSingleton` isn’t recreated because it's already memoized as a singleton in the container.

3. Now, let's create an instance of `ASingleton` - `cnt.use(ASingleton.class)`. This is, I believe, the case you were interested in. 
```
BTransient created
ASingleton created
```
When creating an instance of `ASingleton`, the transient dependency `BTransient` is recreated, but `CSingleton` is reused from the memoized instance.

4. When you try to get `ASingleton` again, you won't see any messages, because no constructor was called. The container returns the memoized instance of `ASingleton`. No constructors are called, and the private fields (`b` and `c`) continue to reference the same instances as when `ASingleton` was first created. So, in this context, you can think that `b` and `c` have become singletons, but if you call `cnt.use(BTransient.class)` you will get a new instance as it is still defined as transient.

In other words, when requesting a singleton that is already memoized in the container, the library won't recreate its dependencies, and you will get the exact unchanged object as it was created for the first time.

Please let me know if that answers your question.

**Comment by angelhdzdev** [comment]
Thank you for the detailed explanation! Appreciate it so much!

And yes, I finally understood why parent class must be transient so that other transient deps can be recreated.
Used an A.I. to reach this conclusion. 

I found this library in my journey to make my own NestJS inspired framework using TypeScript official decorators (which dont include parameter decorators), to avoid being behind the `experimentalDecorators` flag. Now my mini framework is complete!

I used to inject stuff directly in the class in a `props` static field, but your implementation makes more sense and I immediately loved it.

I built `@injectable()` decorator that adds the   `.class = cls.singleton(this)` to all the class constructors when building my controllers and their routes, using default `singleton` and accepts the scope option, also `@singleton`, `@transient`.


```ts
export @controller('auth') 
class AuthController {
  static class = cls.singleton(this, [AuthService])
  
  constructor(private readonly authService: AuthService) {}
  
  @post('sign-in')
  @args(body('email'), body('password'), @session())
  signIn(email: string, password: string, session: UserSession) {
    const user = await this.authService.signIn(email, password)
    session.data.user = {id: user.id}
    const {password:_, ...authUser} = user
    return authUser
  }
}

```

Wishing you success and lots of health. 👋😄
